sdk/python/kfp/_local_client.py

```diff
@@ -0,0 +1,503 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import datetime
+import json
+import logging
+import re
+import subprocess
+from collections import deque
+from typing import Any, Callable, Dict, List, Mapping, Tuple, Union, cast
+
+from . import dsl
+
+
+class _Dag:
+    """DAG stands for Direct Acyclic Graph.
+
+    DAG here is used to decide the order to execute pipeline ops.
+
+    For more information on DAG, please refer to `wiki <https://en.wikipedia.org/wiki/Directed_acyclic_graph>`_.
+
+    """
+
+    def __init__(self, nodes: List[str]) -> None:
+        """
+
+        Args::
+          nodes: List of DAG nodes, each node is identified by an unique name.
+        """
+        self._graph = {node: [] for node in nodes}
+        self._reverse_graph = {node: [] for node in nodes}
+
+    @property
+    def graph(self):
+        return self._graph
+
+    @property
+    def reverse_graph(self):
+        return self._reverse_graph
+
+    def add_edge(self, edge_source: str, edge_target: str) -> None:
+        """Add an edge between DAG nodes.
+
+        Args::
+          edge_source: the source node of the edge
+          edge_target: the target node of the edge
+        """
+        self._graph[edge_source].append(edge_target)
+        self._reverse_graph[edge_target].append(edge_source)
+
+    def get_follows(self, source_node) -> List[str]:
```

Comment:
Are nodes here strings? If so, can we add typehint for that here and elsewhere below?